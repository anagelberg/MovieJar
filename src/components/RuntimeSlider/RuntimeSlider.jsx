
import './RuntimeSlider.scss';
import { useState } from 'react';

function RuntimeSlider({ value, onChange, min, max }) {
    const [perFilled, setPerFilled] = useState((value - min) * 100 / (max - min));

    const handleChange = (event) => {
        onChange(event.target.value);
        setPerFilled((event.target.value - min) * 100 / (max - min));
    };

    return (
        <div className="slider">
            <div className="slider__contents">
                <label className="slider__label" htmlFor='runtime-slider'>Select max runtime (minutes)</label>
                <div className="slider__track">
                    <div
                        className="slider__track-filled"
                        style={{ width: `${perFilled}%` }}
                    >
                    </div>
                </div>
                <input
                    type="range"
                    min={min}
                    max={max}
                    id="runtime-slider"
                    value={value}
                    aria-valuemin={min}
                    aria-valuemax={max}
                    aria-valuenow={value}
                    onChange={handleChange}
                    className="slider__slider"
                />

            </div>
            <div className="slider__value">{value}{value >= 200 && '+'}</div>
        </div>
    );
}

export default RuntimeSlider;
