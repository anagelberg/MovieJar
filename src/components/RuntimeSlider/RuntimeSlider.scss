@use "../../styles/variables" as *;
@use "../../styles/mixins" as *;

$thumb-size: $field-input-height;
$track-height: calc($field-input-height - $spacing-unit-xs);

// replace default thumb circle
@mixin new-thumb {
  -webkit-appearance: none;
  width: $thumb-size;
  height: $thumb-size;
  border-radius: 50%;
  border: 1px solid $secondary;
  background: $primary;
  cursor: pointer;
}

.slider {
  display: flex;
  @include flex-center(
    $justify-content: space-between,
    $align-content: flex-end
  );

  &__contents {
    position: relative;
    width: 85%;
  }

  &__slider {
    appearance: none;
    -webkit-appearance: none;
    position: relative;
    width: 100%;
    height: calc($thumb-size - $spacing-unit-xs);
    border-radius: $border-radius;
    background: transparent;
    outline: none;
    transition: $transition;
    cursor: pointer;
    z-index: 2;

    &::-webkit-slider-thumb {
      @include new-thumb;
    }
    &::-moz-range-thumb {
      @include new-thumb;
    }
  }

  &__track {
    position: absolute;
    width: 100%;
    height: $track-height;
    background: $light-background;
    border-radius: $border-radius;
  }

  &__track-filled {
    position: absolute;
    height: 100%;
    background: $primary;
    border-radius: $border-radius;
    z-index: 1;
  }

  &__label {
    @include form-label;
  }
}
