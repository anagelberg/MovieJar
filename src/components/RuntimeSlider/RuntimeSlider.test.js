import { render, screen, fireEvent } from '@testing-library/react';
import RuntimeSlider from './RuntimeSlider';

describe("RuntimeSlider component", () => {

    it('renders with initial value and calculates the filled percentage correctly', () => {
        const initialValue = 50;
        const min = 0;
        const max = 100;
        const perFilled = ((initialValue - min) * 100) / (max - min);

        const { container } = render(<RuntimeSlider value={initialValue} onChange={() => { }} min={min} max={max} />);

        const sliderElement = screen.getByRole('slider');
        expect(sliderElement.value).toBe(String(initialValue));

        const filledTrack = container.querySelector('.slider__track-filled');
        expect(filledTrack.style.width).toBe(`${perFilled}%`);
    });

    it('updates value and filled percentage on slider change', () => {
        const handleChange = jest.fn();
        const initialValue = 20;
        const newValue = 40;
        const min = 0;
        const max = 100;
        const newPerFilled = ((newValue - min) * 100) / (max - min);

        const { container } = render(<RuntimeSlider value={initialValue} onChange={handleChange} min={min} max={max} />);

        const sliderElement = screen.getByRole('slider');
        fireEvent.change(sliderElement, { target: { value: newValue } });

        expect(handleChange).toHaveBeenCalledWith(String(newValue));

        const filledTrack = container.querySelector('.slider__track-filled');
        expect(filledTrack.style.width).toBe(`${newPerFilled}%`);
    });

    it('displays + sign when value is 200 or more', () => {
        render(<RuntimeSlider value={200} onChange={() => { }} min={0} max={300} />);
        expect(screen.getByText('200+')).toBeInTheDocument();
    });


    it('has correct ARIA attributes for accessibility', () => {

        const initialValue = 50;
        const min = 0;
        const max = 100;

        render(<RuntimeSlider value={initialValue} onChange={() => { }} min={min} max={max} />);

        const sliderElement = screen.getByRole('slider');
        expect(sliderElement.getAttribute('aria-valuemin')).toBe(String(min));
        expect(sliderElement.getAttribute('aria-valuemax')).toBe(String(max));
        expect(sliderElement.getAttribute('aria-valuenow')).toBe(String(initialValue));
    });

});