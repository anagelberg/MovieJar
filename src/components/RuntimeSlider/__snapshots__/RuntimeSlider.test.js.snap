// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuntimeSlider component matches the snapshot 1`] = `
<div>
  <div
    class="slider"
  >
    <div
      class="slider__contents"
    >
      <label
        class="slider__label"
        for="runtime-slider"
      >
        Select max runtime (minutes)
      </label>
      <div
        class="slider__track"
      >
        <div
          class="slider__track-filled"
          style="width: 50%;"
        />
      </div>
      <input
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="50"
        class="slider__slider"
        id="runtime-slider"
        max="100"
        min="0"
        type="range"
        value="50"
      />
    </div>
    <div
      class="slider__value"
    >
      50
    </div>
  </div>
</div>
`;
